src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineControl.xaml.cs

```diff
@@ -0,0 +1,388 @@
+                {
+                    _trackedTextViews.Add(pTextView, wpfTextView);
+                    wpfTextView.Caret.PositionChanged += Caret_PositionChanged;
+                    wpfTextView.TextBuffer.Changed += TextBuffer_Changed;
+                }
+            }
+        }
+
+        int IVsCodeWindowEvents.OnCloseView(IVsTextView pView)
+        {
+            if (_trackedTextViews.TryGetValue(pView, out var view))
+            {
+                view.Caret.PositionChanged -= Caret_PositionChanged;
+                view.TextBuffer.Changed -= TextBuffer_Changed;
+
+                _trackedTextViews.Remove(pView);
+            }
+
+            return VSConstants.S_OK;
+        }
+
+        private void TextBuffer_Changed(object sender, TextContentChangedEventArgs e)
+            => StartGetModelTask();
+
+        // On caret position change, highlight the corresponding symbol node
+        private void Caret_PositionChanged(object sender, CaretPositionChangedEventArgs e)
+        {
+            if (!e.NewPosition.Equals(e.OldPosition))
+                StartHightlightNodeTask();
+        }
+
+        /// <summary>
+        /// Fetches and processes the current document model.
+        /// </summary>
+        private async ValueTask GetModelAsync(CancellationToken cancellationToken)
+        {
+            await TaskScheduler.Default;
+            CodeWindow.GetLastActiveView(out var textView);
+            var activeTextView = EditorAdaptersFactoryService.GetWpfTextView(textView);
+            if (activeTextView is not null)
+            {
+                // Need to be on UI thread to get the file path.
+                await ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+                var filePath = GetFilePath(activeTextView);
+
+                // Ensure we switch to the threadpool before calling DocumentSymbolsRequestAsync.  It ensures
+                // that fetching and processing the document model is not done on the UI thread.
+                await TaskScheduler.Default;
+                var response = await DocumentOutlineHelper.DocumentSymbolsRequestAsync(
+                    activeTextView.TextBuffer, LanguageServiceBroker, filePath, cancellationToken).ConfigureAwait(false);
+
+                if (response is not null)
+                {
+                    var responseBody = response.ToObject<DocumentSymbol[]>();
+                    var documentSymbols = DocumentOutlineHelper.GetNestedDocumentSymbols(responseBody);
+                    LatestFetchedDocumentModel = DocumentOutlineHelper.GetDocumentSymbolModels(documentSymbols);
+
+                    LspSnapshot = activeTextView.TextSnapshot;
+                    LatestFetchedDocumentModelIsNotEmpty = LatestFetchedDocumentModel.Length > 0;
+
+                    StartModelUpdateTask();
+                }
+                else
+                {
+                    LatestFetchedDocumentModelIsNotEmpty = false;
```

Comment:
i'm curious why we need this boolean.  can't we just tell by examining LatestFetchedDocumentModel and seeing if it's empty or not?  what's the difference between those two values (if there is a difference, please doc the fields).

Reply:
We don't need this boolean anymore